src/rdb.c

```diff
@@ -2185,6 +2217,161 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, int dbid, int *error) {
+                        rdbReportCorruptRDB("Hash with dup elements");
+                        dictRelease(dupSearchDict);
+                        decrRefCount(o);
+                        sdsfree(field_dup);
+                        sdsfree(value);
+                        sdsfree(field);
+                        return NULL;
+                    }
+                }
+
+                /* check if the values can be saved to listpack (or should convert to dict encoding) */
+                if (sdslen(field) > server.hash_max_listpack_value ||
+                    sdslen(value) > server.hash_max_listpack_value ||
+                    !lpSafeToAdd(((listpackEx*)o->ptr)->lp, sdslen(field) + sdslen(value) + lpEntrySizeInteger(expire)))
+                {
+                    /* convert to hash */
+                    hashTypeConvert(o, OBJ_ENCODING_HT, hexpires);
+
+                    if (len > DICT_HT_INITIAL_SIZE) { /* TODO: this is NOT the original len, but this is also the case for simple hash, is this a bug? */
+                        if (dictTryExpand(o->ptr, len) != DICT_OK) {
+                            rdbReportCorruptRDB("OOM in dictTryExpand %llu", (unsigned long long)len);
+                            decrRefCount(o);
+                            if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+                            sdsfree(value);
+                            sdsfree(field);
+                            return NULL;
+                        }
+                    }
+
+                    /* don't add the values to the new hash: the next if will catch and the values will be added there */
+                } else {
+                    listpackExAddNew(o, field, value, expire);
+                    sdsfree(field);
+                    sdsfree(value);
+                }
+            }
+
+            if (o->encoding == OBJ_ENCODING_HT) {
+                /* WA for check-rdb mode, when there's no DB so can't attach expired items to ebuckets,
+                 * or when no expiry was not set for this field */
+                if ((db == NULL) || (expire == 0)) {
+                    hashTypeSet(db, o, field, value, 0);
+                } else {
+                    if (hashTypeSetExRdb(db, o, field, value, expire) != C_OK) {
+                        serverLog(LL_WARNING, "failed adding hash field %s to key %s", field, key);
+                        decrRefCount(o);
+                        if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+                        sdsfree(value);
+                        sdsfree(field);
+                        return NULL;
+                    }
+                }
+                sdsfree(field);
+                sdsfree(value);
+            }
+        }
+
+        if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+
+        /* check for empty key (if all fields were expired) */
+        if (hashTypeLength(o, 0) == 0) {
+            decrRefCount(o);
+            goto emptykey;
```

Comment:
if we exit here and o->encoding is hashtable, does it mean that db->hexpires still retain the robj?

Reply:
Presumably no, the new object is only added to `db->hexpires' in the following `if` statement, so if the function returns here it was not yet added (and therefore should not be removed, obviously)